File: Generics/app/src/main/java/com/example/generics/MainActivity.java
package com.example.generics;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.AppCompatTextView;

import android.os.Bundle;
import android.view.View;
import android.widget.TextView;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        AppleArrayList appleArrayList = new AppleArrayList();
        appleArrayList.add(new Apple());

//        ArrayList<? extends View> fruitList = new ArrayList<TextView>();

        ArrayList<? super AppCompatTextView> fruitList = new ArrayList<TextView>();

        List<String> list = new ArrayList<String>(){};
    }

}

File: CustomViewLayoutSize/app/src/main/java/com/example/customviewlayoutsize/MainActivity.kt
package com.example.customviewlayoutsize

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle

/**
 * "自定义布局"
 *
 * 继承已有的View，简单改写他们的尺寸：重写onMeasure()
 * SquareImageView
 *
 * 对自定义View完全进行自定义尺寸计算：重写onMeasure()
 * CircleImage
 *
 * 自定义Layout：重写onMeasure()和onLayout()
 */
class MainActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
    }
}